### PR TITLE
fix: OnEvent is not propagating data in some scenarios with postgres

### PR DESCRIPTION
**Pull request by duonganhthu43**
related [issue](https://github.com/getdozer/dozer/issues/625)

### CODE REVIEW

File: dozer-api/src/grpc/internal/internal_pipeline_server.rs [hunk 1]
```diff
@@ -6,45 +6,71 @@ use crate::grpc::internal_grpc::{
...
+            receiver: rx1,
+        }
+    }
+
+    fn setup_broad_cast_channel(
+        tx: Sender<PipelineResponse>,
+        receiver: Receiver<PipelineResponse>,
+    ) {
+        while let Ok(event_response) = receiver.recv() {
+            let _ = tx.send(event_response);
```
Comment (chubei) [inline dozer-api/src/grpc/internal/internal_pipeline_server.rs:37]: Maybe print a log if it's an `Err`?

**Review by chubei — COMMENTED** [review]
I'm wondering if `broadcast` is the root cause at all. We're mixing `crossbeam` channels and `tokio` channels, which is bad in tokio runtime because `crossbeam` channels are not async-aware. Two futures can run in the same thread while one is sending and another is receiving. I think we need to remove all the crossbeam channels in async code.

File: dozer-api/src/grpc/internal/internal_pipeline_server.rs (hunk 1 above)
Comment (duonganhthu43) [reply dozer-api/src/grpc/internal/internal_pipeline_server.rs:37]: yah, i want to get rid of `crossbeam` also,  but it using across everywhere in `dozer-orchestrator` gonna have a huge affect on that.

`broadcast` is for solving issue multiple clients subscribe to that method. Just found out that bug: when a client is subscribe, later client subscribe won't receive any message

**Comment by chubei** [comment]
> when a client is subscribe, later client subscribe won't receive any message

I'm a little confused. I thought `InternalPipelineServer` is the only one subscribing to pipeline events?

**Comment by duonganhthu43** [comment]
> > when a client is subscribe, later client subscribe won't receive any message
> 
> I'm a little confused. I thought `InternalPipelineServer` is the only one subscribing to pipeline events?
 
 our `API` + `Admin` is the client of `Internal Server`

**Comment by chubei** [comment]
> our `API` + `Admin` is the client of `Internal Server`

Yes, but the `receiver` in `InternalPipelineServer` is for `InternalPipelineServer` subscribing to the pipeline, not the clients subscribing to it, right?

**Comment by duonganhthu43** [comment]
> > our `API` + `Admin` is the client of `Internal Server`
> 
> Yes, but the `receiver` in `InternalPipelineServer` is for `InternalPipelineServer` subscribing to the pipeline, not the clients subscribing to it, right?

Yup, it is

**Comment by chubei** [comment]
> 

Oh so you just set up the extra channel to go from crossbeam to tokio right? Not for the `broadcast`.

I'm trying to understand what fixes the problem here.

**Comment by chubei** [comment]
I decided not to remove `float_value` because my analysis of the bug case was wrong.

We don't have a round trip of `dozer_types::types::Field` to `prost_refect::Value` and then to `dozer_types::types::Field`. What we have are as following:

- `dozer_types::types::Record` -> `grpc::types::Recrod`. Used in common service `query` method and `CacheSink`. Implemented in `dozer_api::grpc::types_helper`.
- `grpc::types::Record` -> `prost_reflect::DynamicMessage`. Used in typed service `on_event` method. Implemented in `dozer_api::grpc::typed::helper`.
- `dozer_types::types::Record` -> `prost_reflect::DynamicMessage`. Used in typed service `query` method. Implemented in `dozer_api::grpc::typed::helper`.

The reason that we need to use different conversion paths for `query` and `on_event` is that, for `query`, API server is directly accessing the cache, getting `dozer_types::types::Record`; for `on_event`, events first get converted to `dozer_api::grpc::types` types and then sent from app server to API server.

**Review by chubei — APPROVED** [review]

File: dozer-api/src/grpc/internal/internal_pipeline_server.rs [hunk 2]
```diff
@@ -6,45 +6,83 @@ use crate::grpc::internal_grpc::{
...
 
 #[tonic::async_trait]
 impl InternalPipelineService for InternalPipelineServer {
     type StreamPipelineRequestStream = ResponseStream;
-
     async fn stream_pipeline_request(
         &self,
         _request: tonic::Request<PipelineRequest>,
     ) -> Result<Response<ResponseStream>, Status> {
         let (tx, rx) = tokio::sync::mpsc::channel(1000);
```
Comment (v3g42) [inline dozer-api/src/grpc/internal/internal_pipeline_server.rs:62]: This is the problem I believe. We should have channels that we use across pipeline to websocket comparable in terms of capacity. We probably need to give internal more capacity than what pipeline is configured. Ideally this should be adjustable through configuration @chubei @duonganhthu43 what do you think ? 

When i tested locally with higher capacity, eth messages were not dropped.

**Comment by v3g42** [comment]
@duonganhthu43 @chubei can you summarize what  fixes the problem here ?

File: dozer-api/src/grpc/internal/internal_pipeline_server.rs (hunk 2 above)
Comment (chubei) [reply dozer-api/src/grpc/internal/internal_pipeline_server.rs:62]: Interesting. I thought we wound block on channel becoming full. I think we should find why channel capacity affects the program behaviour becuaes otherwise backpressure won't be working properly.

Comment (duonganhthu43) [reply dozer-api/src/grpc/internal/internal_pipeline_server.rs:62]: we could try `unbounded_channel` if capacity is the problem

**Comment by chubei** [comment]
Here's my understanding:

The internal server spawns a new future for every client listening to the event stream. Let's call these futures L1, L2...

The internal server itself is a spawned future. Let's call it S1.

S1 communicates with L1 through a channel.

Problem is that we used a `crossbeam::channel` for this communication.

When the channel is full, S1 blocks until L1 receives an item from the channel. However, if it so happens that S1 and L1 are running on the same thread, the blocking of the sending of S1 implies L1 can never get the opportunity to execute receiving.

The same happens when channel is empty and L1 gets blocked on receiving while S1 is on the same thread waiting for L1 to `await`.

The solution is to switch from `crossbeam::channel` to `tokio::channel` because tokio channels are async-aware and yields execution flow when sending or receiving blocks.

**Comment by duonganhthu43** [comment]
> @duonganhthu43 @chubei can you summarize what fixes the problem here ?

few issues found out: 

- Previously, the Internal server only served 1 client subscribe => so the next client subscribed won't receive any event 
- One minor bug in casting value from `Pipeline event` to Proto type `on_event`, double and float converting => Our Postgres sample have this case 
- A delay async issue when streaming event back to client => some client not received event immediately but wait for next event, then received old + new event => That's why Ethereum with so much event always works, Postgres you have to produce at least 2 events to received it

**Comment by v3g42** [comment]
@chubei @duonganhthu43 Try out increasing the capacity. But i suspect it is that. As I was able to run ethereum example with no issues with the same config ? 

@chubei should we merge https://github.com/getdozer/dozer/pull/647 onto this PR and test together just in case? 

@duonganhthu43 let me know if you are succesfully able to run the sample. 

Can you also track taking the capacity of the changes through config as a separate issue ?

**Comment by chubei** [comment]
#647 won't have effect on this issue. I think we can leave it there. Or merge it before this one.

**Comment by duonganhthu43** [comment]
> @duonganhthu43 let me know if you are succesfully able to run the sample.

yah, able to run now in this branch

File: dozer-api/src/grpc/internal/internal_pipeline_server.rs (hunk 2 above)
Comment (duonganhthu43) [reply dozer-api/src/grpc/internal/internal_pipeline_server.rs:65]: will create issue with channel capacity config
